src/main/scala/is/hail/expr/Type.scala

```diff
@@ -2277,17 +2288,22 @@ final case class TStruct(fields: IndexedSeq[Field], override val required: Boole
   }
 
   def isFieldDefined(region: MemoryBuffer, offset: Long, fieldIdx: Int): Boolean =
-    !region.loadBit(offset, fieldIdx)
+    fields(fieldIdx).typ.required || !region.loadBit(offset, missingIdx(fieldIdx))
```

Comment:
use `fieldType(fieldIdx)`.

Reply:
You do this in a number of places.  You might also consider adding `isFieldRequired(fieldIdx: Int): Boolean`.